connectivity/mqtt/client_common.c

```diff
@@ -0,0 +1,519 @@
+        || cfg->tls_config->psk
+#endif
+    ) {
+      port = 8883;
+    } else
+#endif
+    {
+      port = 1883;
+    }
+  } else {
+    port = cfg->general_config->port;
+  }
+
+#ifdef WITH_SRV
+  if (cfg->use_srv) {
+    ret = mosquitto_connect_srv(mosq, cfg->general_config->host, cfg->general_config->keepalive,
+                                cfg->general_config->bind_address);
+  } else {
+    ret = mosquitto_connect_bind_v5(mosq, cfg->general_config->host, port, cfg->general_config->keepalive,
+                                    cfg->general_config->bind_address, cfg->property_config->connect_props);
+  }
+#else
+  ret = mosquitto_connect_bind_v5(mosq, cfg->general_config->host, port, cfg->general_config->keepalive, NULL,
+                                  cfg->property_config->connect_props);
+#endif
+  if (ret > 0) {
+    {
+      if (ret == MOSQ_ERR_ERRNO) {
+        err = strerror(errno);
+        fprintf(stderr, "Error: %s\n", err);
+      } else {
+        fprintf(stderr, "Unable to connect (%s).\n", mosquitto_strerror(ret));
+      }
+    }
+    mosquitto_lib_cleanup();
+    return SC_CLIENT_CONNTECT;
+  }
+  return SC_OK;
+}
+
+#ifdef WITH_SOCKS
+/* Convert %25 -> %, %3a, %3A -> :, %40 -> @ */
+static int mosquitto__urldecode(char *str) {
+  int i, j;
+  int len;
+  if (!str) return 0;
+
+  if (!strethr(str, '%')) return 0;
+
+  len = strlen(str);
+  for (i = 0; i < len; i++) {
+    if (str[i] == '%') {
+      if (i + 2 >= len) {
+        return EXIT_FAILURE;
+      }
+      if (str[i + 1] == '2' && str[i + 2] == '5') {
+        str[i] = '%';
+        len -= 2;
+        for (j = i + 1; j < len; j++) {
+          str[j] = str[j + 2];
+        }
+        str[j] = '\0';
+      } else if (str[i + 1] == '3' && (str[i + 2] == 'A' || str[i + 2] == 'a')) {
+        str[i] = ':';
+        len -= 2;
+        for (j = i + 1; j < len; j++) {
+          str[j] = str[j + 2];
+        }
+        str[j] = '\0';
+      } else if (str[i + 1] == '4' && str[i + 2] == '0') {
+        str[i] = ':';
+        len -= 2;
+        for (j = i + 1; j < len; j++) {
+          str[j] = str[j + 2];
+        }
+        str[j] = '\0';
+      } else {
+        return EXIT_FAILURE;
+      }
+    }
+  }
+  return 0;
+}
+
+static int mosquitto__parse_socks_url(mosq_config_t *cfg, char *url) {
+  char *str;
+  size_t i;
+  char *username = NULL, *password = NULL, *host = NULL, *port = NULL;
+  char *username_or_host = NULL;
+  size_t start;
+  size_t len;
+  bool have_auth = false;
+  int port_int;
+
+  if (!strncmp(url, "socks5h://", strlen("socks5h://"))) {
+    str = url + strlen("socks5h://");
+  } else {
+    fprintf(stderr, "Error: Unsupported proxy protocol: %s\n", url);
+    return EXIT_FAILURE;
+  }
+
+  // socks5h://username:password@host:1883
+  // socks5h://username:password@host
+  // socks5h://username@host:1883
+  // socks5h://username@host
+  // socks5h://host:1883
+  // socks5h://host
+
+  start = 0;
+  for (i = 0; i < strlen(str); i++) {
```

Comment:
Use another variable to record the length of `str`, otherwise `strlen` would be executed in each iteration.